### ISSUE TITLE
[BUG] Invalid weightObjRef causes failure handler failures

### ISSUE THREAD

**Issue by sriumcp** [iter8-tools/iter8#644]
**Describe the bug**
If weightObjRef in experiment is invalid (example missing in a progressive deployment A/B testing experiment), it invokes the finish handler, which fails because the controller is yet to set versionRecommendedForPromotion.

**To Reproduce**
Remove the weightObjRefs in Knative Quickstart tutorial.

**Expected behavior**
If there is a failure (e.g., weightObjRef in experiment is invalid), finish handler should not be invoked.

**Comment by kalantar** [comment]
The finish handler is being called because there is a failure. In the past, we called this a failure handler. However, we determined that we should call the same handler in both cases.  I think I am ok with the suggested change. However, we should consider renaming the action type to be `promotion` instead of finish. This would make clear to the user when it is expected to be called. This could be a non-breaking change; we just deprecate the use of `finish`.

Alternatively, a well written handler should be robust. Given greater visibility to the experiment t could have figured this out and taken appropriate action. I don't like the automatic interpolation.

**Comment by sriumcp** [comment]
Separating `promotion` from `failure` seems like a good idea. However, I do remember a separate discussion and agreement on not calling the failure handler when start handler failed (which I believe is now implemented). The same principle should apply for other failures.

I am curious about why you don't like automatic interpolation, and what the alternative approaches are to promotion.

**Comment by sriumcp** [comment]
```yaml
  conditions:
  - lastTransitionTime: "2021-05-03T14:51:32Z"
    message: Failure handler failed
    reason: ExperimentCompleted
    status: "True"
    type: Completed
  - lastTransitionTime: "2021-05-03T14:43:57Z"
    status: "False"
    type: Failed
  - lastTransitionTime: "2021-05-03T14:43:57Z"
    message: ""
    reason: TargetAcquired
    status: "True"
    type: TargetAcquired
```

Here is another example of experiment status output... too many things are off with this output, namely:
1. There is no concept of a `failure` handler at the moment (`start`, `loop` and `finish`) are the only three documented handlers.
2. It is unclear why Iter8 attempted to call the failure handler in the first place.
3. Taking for granted that failure handler = finish handler, it is unclear why finish handler failed.

**Comment by kalantar** [comment]
> 1. There is no concept of a `failure` handler at the moment

Agreed. I searched the etc3 code for this and it is no longer present. It was fixed in https://github.com/iter8-tools/etc3/pull/200. I just created a [new revision](https://github.com/iter8-tools/etc3/releases/tag/v0.1.22) we will need to adopt.

In this case, it could have been any handler that failed. The above fix will get the right type of handler.

> it is unclear why finish handler failed

etc3 does not have visibility into why a handler failed; to determine this we would have to scrap the job output. However, we could add instructions for seeing the log to the message:

    XXX handler failed. To see the logs: kubectl -n iter8-system log job/EXPERIMENT-XXX
